test_lbaas.py

```diff
@@ -0,0 +1,625 @@
+
+    Note: The current connection count is reset on any configuration changes to
+          the load balancer. Still running connections initiated before the
+          configuration change are not considered for selecting the backend.
+    """
+
+    # This test uses a function scoped prober because we are going to start
+    # long running downloads which should not impact other tests.
+    prober = server
+
+    # Create a load balancer setup with 2 backends on a private network
+    load_balancer, listener, pool, backends, private_network = \
+        create_load_balancer_scenario(
+            num_backends=2,
+            algorithm='least_connections',
+        )
+
+    # start a persistent endless download of random data to "block" one backend
+    backend_endless = start_persistent_download(prober, load_balancer,
+                                                backends)
+
+    # Verify requests go to the other backend as it has less active
+    # connections
+    for i in range(10):
+        backend_hit = load_balancer.get_url(prober, url='/hostname')
+
+        assert backend_hit != backend_endless.name
+
+
+@pytest.mark.parametrize('health_monitor',
+                         [{'type': 'ping',
+                           'http': None},
+                          {'type': 'tcp',
+                           'http': None},
+                          {'type': 'http',
+                           'http': {'host': 'www.example.com'}},
+                          {'type': 'https',
+                           'http': {'version': '1.0'}},
+                          {'type': 'tls-hello',
+                           'http': None},
+                          ],
+                         ids=['ping', 'tcp', 'http', 'https', 'tls-helo'])
+def test_health_monitors(prober, create_load_balancer_scenario,
+                         health_monitor):
+    """ Different health monitoring methods can be used to verify pool member
+    availability:
+
+    * ICMP ping health monitors check pool member availability by sending ICMP
+      echo requests to the pool member IP address
+    * TCP health monitors check the availability of TCP port on a pool member
+    * HTTP(S) health monitors check the response code of a HTTP(S) URL
+    * TLS_HELLO health monitors check that a pool member answers by initiating
+      a TLS handshake on a TCP port
+
+    """
+
+    # Configure SSL backend for SSL health checks
+    ssl = health_monitor['type'] in ('https', 'tls-hello') and True or False
+
+    # Create a load balancer setup with 1 backend on a private network
+    load_balancer, listener, pool, (backend, ), private_network = \
+        create_load_balancer_scenario(
+            num_backends=1,
+            health_monitor=health_monitor,
+            ssl=ssl,
+    )
+
+    # Test function to assert the desired load balancer and health monitor
+    # status
+    def assert_load_balancer_and_monitor_status(load_balancer_status,
```

Comment:
I think the function signature (name and variables) is overly long here. Maybe use `lb` instead of `load_balancer`.

Reply:
As it's only a nested function I renamed it to `assert_status`.